lib/Extension/LanguageServerCodeTransform/LspCommand/GenerateDecoratorCommand.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace Phpactor\Extension\LanguageServerCodeTransform\LspCommand;
+
+use Amp\Promise;
+use Phpactor\CodeBuilder\Domain\Builder\SourceCodeBuilder;
+use Phpactor\Extension\LanguageServerBridge\Converter\TextEditConverter;
+use Phpactor\LanguageServerProtocol\Command;
+use Phpactor\LanguageServerProtocol\WorkspaceEdit;
+use Phpactor\LanguageServer\Core\Server\ClientApi;
+use Phpactor\CodeTransform\Domain\SourceCode;
+use Phpactor\LanguageServer\Core\Workspace\Workspace;
+use Phpactor\TextDocument\TextEdits;
+use Phpactor\WorseReflection\Reflector;
+use Phpactor\CodeBuilder\Domain\Builder\MethodBuilder;
+use Phpactor\WorseReflection\Core\Reflection\ReflectionMethod;
+use Phpactor\CodeBuilder\Domain\Prototype\Visibility;
+
+class GenerateDecoratorCommand extends Command
+{
+    public const NAME = 'generate_decorator';
+
+    private ClientApi $clientApi;
+
+    private Workspace $workspace;
+
+    private Reflector $reflector;
+
+    public function __construct(
+        ClientApi $clientApi,
+        Workspace $workspace,
+        Reflector $reflector
+    ) {
+        $this->clientApi = $clientApi;
+        $this->workspace = $workspace;
+        $this->reflector = $reflector;
+    }
+
+    /**
+     * @return Promise<ApplyWorkspaceEditResponse>
+     */
+    public function __invoke(string $uri, string $interfaceFQN): Promise
+    {
+        $textDocument = $this->workspace->get($uri);
+        $source = SourceCode::fromStringAndPath($textDocument->text, $textDocument->uri);
+
+        $textEdits = $this->getTextEdits($source, $interfaceFQN);
+
+        return $this->clientApi->workspace()->applyEdit(new WorkspaceEdit([
+            $uri => TextEditConverter::toLspTextEdits($textEdits, $textDocument->text)
+        ]), 'Generate decoration');
+    }
+
+    private function getTextEdits(SourceCode $source, string $interfaceFQN): TextEdits
+    {
+        $class = $this->reflector->reflectClassesIn($source)->classes()->first();
+
+        $builder = SourceCodeBuilder::create();
+        $builder->namespace($class->name()->namespace());
+        $classBuilder = $builder->class($class->name()->short());
+
+        $classBuilder->property('inner')->visibility(Visibility::PRIVATE)->type($interfaceFQN);
+        $constructor = $classBuilder->method('__construct');
+
+        $constructor->parameter('inner')->type($interfaceFQN);
+        $constructor->body()->line('$this->inner = $inner');
+
+        $interface = $this->reflector->reflectInterface($interfaceFQN);
+        foreach ($interface->methods() as $interfaceMethod) {
+            $method = $classBuilder->method($interfaceMethod->name());
+
+            $method->returnType($interfaceMethod->returnType());
+            $method->visibility($interfaceMethod->visibility());
+
+            $this->attachParameters($method, $interfaceMethod);
+
+            $method->body()->line($this->generateMethodBody($interfaceMethod));
+        }
+
+        return [];
```

Comment:
I think I have done most of it but there is one question. I have generates all the methods and the new property inside the decorator class but how to I extract a `TextEdits` out of them?

Reply:
If using the "code builder" thing:

```
        return $this->updater->textEditsFor($sourceCodeBuilder->build(), Code::fromString((string) $source));
```

to transform to _phpactor_ text edits. then need to convert to LSP edits in the code action provider (the following code is in `.../LspCommand/TransformCommand` FWIW)

```
        return $this->clientApi->workspace()->applyEdit(new WorkspaceEdit([
            $uri => TextEditConverter::toLspTextEdits($textEdits, $textDocument->text)
        ]), 'Apply source code transformation');
````

Reply:
Well then I should be done.